src/bindings/js/node/src/infer_request.cpp

```diff
@@ -190,3 +197,82 @@ void InferRequestWrap::infer(const Napi::Object& inputs) {
 Napi::Value InferRequestWrap::get_compiled_model(const Napi::CallbackInfo& info) {
     return CompiledModelWrap::Wrap(info.Env(), _infer_request.get_compiled_model());
 }
+
+struct TsfnContext {
+    TsfnContext(Napi::Env env) : deferred(Napi::Promise::Deferred::New(env)){};
+
+    Napi::Promise::Deferred deferred;
+    ov::InferRequest* _context_ir;
+    std::thread nativeThread;
+
+    Napi::ThreadSafeFunction tsfn;
+    std::vector<ov::Tensor> input_tensors;
+    std::map<std::string, ov::Tensor> result;
+};
+
+void FinalizerCallback(Napi::Env env, void* finalizeData, TsfnContext* context) {
+    context->nativeThread.join();
+    delete context;
+};
+
+void performInferenceThread(TsfnContext* data) {
+    infer_mutex.lock();
```

Comment:
for locking check if `std::lock_guard` can be used it keeps lock until guard will be removed.

Reply:
or do like:
```cpp
void my_func(){
    {
        std::lock_guard my_lock;
        critical_section();
    }
    
    not_ciritical_section();
}
```

and lock will be applied in code block only and be released before end of function. Code block can make suggestion to make function extraction in future.

Reply:
> But here I can release lock before the end of the function and its scope

`std::unique_lock` is suited for that